File: 2023/inCpp/1st_day/Part2.cpp
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

struct num_spell {
    string spell;
    char number;
};

const num_spell spell_to_number[] = {
    {"one", '1'}, {"two", '2'}, {"three", '3'}, {"four", '4'}, {"five", '5'},
    {"six", '6'}, {"seven", '7'}, {"eight", '8'}, {"nine", '9'}
};

string replace_spell(const string& str) {
    string result = str;

    for (const num_spell& mapping : spell_to_number) {
        size_t pos = result.find(mapping.spell);

        while (pos != string::npos) {
            // Replace the second and subsequent characters with the digit
            result[pos + 1] = mapping.number;

            // Move the position to the next character after the replaced digit
            pos = result.find(mapping.spell, pos + 1);
        }
    }

    return result;
}



int main() {
    
    ifstream inputFile;

    inputFile.open("input.txt");

    int sum = 0;
    string line;
    while (getline(inputFile, line)) {
        string num_string = "";
        string modified_str = replace_spell(line);
        // cout << modified_str << endl;
        for (int i = 0; i < modified_str.length(); i++) {
            if (modified_str[i] >= '0' && modified_str[i] <= '9') {
                num_string += modified_str[i];
                break;
            }
        }
        for (int i = modified_str.length() - 1; i >= 0; i--) {
            if (modified_str[i] >= '0' && modified_str[i] <= '9') {
                num_string += modified_str[i];
                break;
            }
        }
        if (!num_string.empty()) {
            // cout << num_string << endl;
            sum += stoi(num_string);
        }
    }

    cout << "Sum = " << sum << endl;
    // Close the file
    inputFile.close();

    return 0;
}

File: 2023/inPy/1st_day/Part2.py
spell_to_number = {
    "one": '1', "two": '2', "three": '3', "four": '4', "five": '5',
    "six": '6', "seven": '7', "eight": '8', "nine": '9'
}

def replace_spell(input_str):
    result = list(input_str)

    for spell, number in spell_to_number.items():
        pos = input_str.find(spell)

        while pos != -1:
            # Replace the second and subsequent characters with the digit
            result[pos + 1] = number

            # Move the position to the next character after the replaced digit
            pos = input_str.find(spell, pos + 1)

    return result


sum_result = 0

with open("1st_day/input.txt", "r") as inputFile:
    for line in inputFile:
        num_string = ""
        line = line.strip()
        replaced_line = replace_spell(line)

        for char in replaced_line:
            if char.isdigit():
                num_string += char
                break

        for char in reversed(replaced_line):
            if char.isdigit():
                num_string += char
                break

        if num_string:
            sum_result += int(num_string)

print("Sum =", sum_result)
